lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -44,29 +53,57 @@ void CTreeShapFeatureImportance::shap(const TRowRef& row, TShapWriter writer) {
         return;
     }
 
+    using TTreeShapVec = std::vector<std::function<void(const TTree&)>>;
+
     auto encodedRow{m_Encoder->encode(row)};
-    m_ShapValues.assign(m_Encoder->numberInputColumns(),
-                        las::zero((*m_Forest)[0][0].value()));
-    for (const auto& tree : *m_Forest) {
-        this->shapRecursive(tree, encodedRow, 0, 1.0, 1.0, -1,
-                            CSplitPath{m_PathStorage.begin(), m_ScaleStorage.begin()},
-                            0, m_ShapValues);
+
+    if (m_PerThreadShapValues.size() == 1) {
+        m_ReducedShapValues.assign(m_Encoder->numberInputColumns(),
+                                   las::zero((*m_Forest)[0][0].value()));
+        for (const auto& tree : *m_Forest) {
+            this->shapRecursive(
+                tree, encodedRow, 0, 1.0, 1.0, -1,
+                CSplitPath{m_PathStorage[0].begin(), m_ScaleStorage[0].begin()},
+                0, m_ReducedShapValues);
+        }
+    } else {
+        TTreeShapVec computeTreeShap;
+        computeTreeShap.reserve(m_PerThreadShapValues.size());
+        for (std::size_t i = 0; i < m_PerThreadShapValues.size(); ++i) {
+            m_PerThreadShapValues[i].assign(m_Encoder->numberInputColumns(),
+                                            las::zero((*m_Forest)[0][0].value()));
+            computeTreeShap.push_back([&encodedRow, i, this](const TTree& tree) {
+                this->shapRecursive(tree, encodedRow, 0, 1.0, 1.0, -1,
+                                    CSplitPath{m_PathStorage[i].begin(),
+                                               m_ScaleStorage[i].begin()},
+                                    0, m_PerThreadShapValues[i]);
+            });
+        }
+
+        core::parallel_for_each(m_Forest->begin(), m_Forest->size(), computeTreeShap);
+
+        m_ReducedShapValues = m_PerThreadShapValues[0];
+        for (std::size_t i = 1; i < m_PerThreadShapValues.size(); ++i) {
```

Comment:
Maybe add a note here that it can be replaces with `std::reduce` and parallel execution once we have C++17 support?

Reply:
I don't think this can go until the standard introduces executors, because we definitely want these to be executed in a thread pool rather than have to spawn threads. So this will have to stay until we are able to adopt C++23 based on current plans. Given how far away this is, I'm not sure it is worth a comment at the moment.